jmclient/jmclient/payjoin.py

```diff
@@ -0,0 +1,503 @@
+        Headers({"Content-Type": ['text/plain']}),
+        bodyProducer=body)
+
+    d.addCallback(receive_payjoin_proposal_from_server, manager)
+    # note that the errback (here "noresponse") is *not* triggered
+    # by a server rejection (which is accompanied by a non-200
+    # status code returned), but by failure to communicate.
+    def noResponse(failure):
+        failure.trap(ResponseFailed)
+        log.error(failure.value.reasons[0].getTraceback())
+        reactor.stop()
+    d.addErrback(noResponse)
+    return (True, None)
+
+def fallback_nonpayjoin_broadcast(manager, err):
+    assert isinstance(manager, JMPayjoinManager)
+    log.warn("Payjoin did not succeed, falling back to non-payjoin payment.")
+    log.warn("Error message was: " + err.decode("utf-8"))
+    original_tx = manager.initial_psbt.extract_transaction()
+    if not jm_single().bc_interface.pushtx(original_tx.serialize()):
+        log.error("Unable to broadcast original payment. The payment is NOT made.")
+    log.info("We paid without coinjoin. Transaction: ")
+    log.info(btc.hrt(original_tx))
+    reactor.stop()
+
+def receive_payjoin_proposal_from_server(response, manager):
+    assert isinstance(manager, JMPayjoinManager)
+
+    # if the response code is not 200 OK, we must assume payjoin
+    # attempt has failed, and revert to standard payment.
+    if int(response.code) != 200:
+        fallback_nonpayjoin_broadcast(manager, err=response.phrase)
+        return
+    # for debugging; will be removed in future:
+    log.debug("Response headers:")
+    log.debug(pformat(list(response.headers.getAllRawHeaders())))
+    # no attempt at chunking or handling incrementally is needed
+    # here. The body should be a byte string containing the
+    # new PSBT.
+    d = readBody(response)
+    d.addCallback(process_payjoin_proposal_from_server, manager)
+
+def process_payjoin_proposal_from_server(response_body, manager):
+    assert isinstance(manager, JMPayjoinManager)
+    try:
+        payjoin_proposal_psbt = \
+            btc.PartiallySignedTransaction.from_base64(response_body)
+    except Exception as e:
+        log.error("Payjoin tx from server could not be parsed: " + repr(e))
+        fallback_nonpayjoin_broadcast(manager, err="Server sent invalid psbt")
+        return
+
+    log.debug("Receiver sent us this PSBT: ")
+    log.debug(manager.wallet_service.hr_psbt(payjoin_proposal_psbt))
+    # we need to add back in our utxo information to the received psbt,
+    # since the servers remove it (not sure why?)
+    for i, inp in enumerate(payjoin_proposal_psbt.unsigned_tx.vin):
+        for j, inp2 in enumerate(manager.initial_psbt.unsigned_tx.vin):
+                    if (inp.prevout.hash, inp.prevout.n) == (
+                        inp2.prevout.hash, inp2.prevout.n):
+                        payjoin_proposal_psbt.inputs[i].set_utxo(
+                            manager.initial_psbt.inputs[j].utxo, None)
```

Comment:
Here you _could_ provide unsigned_tx by doing
`payjoin_proposal_psbt.inputs[i].set_utxo(manager.initial_psbt.inputs[j].utxo, payjoin_proposal_psbt.unsigned_tx)`

I could add psbt-level `psbt.set_utxo(utxo, index=i)` which would take an index to update the input and would use `psbt.unsigned_tx` internally, but I'm not sure if this would be that much of an improvement or less error-prone.

Reply:
I want to keep the processing as our-inputs-segwit-only and decided specifically not to address this "threat" vector, at least for now. The payjoin BIP is still in heavy discussion so there is little possibility that this version of the payjoin functionality will be final, my feeling was: if it turns out once that BIP is finalized that specific provision is made (e.g. requiring nonwitness-utxo style utxo field), then I will change here at that point, otherwise I will prefer segwit-exclusive processing.

As for whether you should change it to automatically include nonwitness/unsigned tx,  well perhaps, yes; I think what I noticed while using the lib is that some functionality is at the lower PSBT_Input level that I would have expected to be exposed at PartiallySignedTransaction level, I guess this is one of those. The internal structure of PartiallySignedTransaction should probably not be a matter for callers.

Reply:
> The internal structure of PartiallySignedTransaction should probably not be a matter for callers.

OK on reflection that is a very debatable statement :) Perhaps it's more just, there are ways users could be provided convenience access at the PartiallySignedTransaction level itself, that seem logic, this is perhaps one such.

Reply:
Not a huge deal of course, but using that new function as of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/c7e7a083591da1235c7bb4093593607b069258b0